Canvas/SubmitAssignment/SubmitAssignmentPresenter.swift

```diff
@@ -0,0 +1,120 @@
+//
+// Copyright (C) 2019-present Instructure, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3 of the License.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+import Core
+import CoreData
+
+protocol SubmitAssignmentView: class {
+    func update()
+}
+
+let group = "group.com.instructure.icanvas"
+
+class SubmitAssignmentPresenter {
+    let env: AppEnvironment
+    weak var view: SubmitAssignmentView?
+
+    private(set) var course: Course? {
+        didSet {
+            assignment = nil
+            if let course = course {
+                assignments = env.subscribe(GetAssignments(courseID: course.id)) { [weak self] in
+                    self?.assignment = self?.assignment ?? self?.assignments?.first
+                }
+                assignments?.refresh()
+            }
+            view?.update()
+        }
+    }
+    private(set) var assignment: Assignment? {
+        didSet { view?.update() }
+    }
+
+    var assignments: Store<GetAssignments>?
+    lazy var courses: Store<GetCourses>? = env.subscribe(GetCourses(showFavorites: false, perPage: 10)) { [weak self] in
+        self?.course = self?.course ?? self?.courses?.first
+    }
+
+    var isContentValid: Bool {
+        return course != nil && assignment != nil
+    }
+
+    init?() {
+        Keychain.config = KeychainConfig(accessGroup: group)
+        guard let session = Keychain.mostRecentSession else { return nil }
+        env = AppEnvironment()
+        env.userDidLogin(session: session)
+    }
+
+    func viewIsReady() {
+        courses?.refresh()
+    }
+
+    func select(course: Course) {
+        self.course = course
+    }
+
+    func select(assignment: Assignment) {
+        self.assignment = assignment
+    }
+
+    func submit(items: [NSExtensionItem], callback: @escaping (Error?) -> Void) {
+        guard let assignment = assignment else { return }
+        let uploadContext = FileUploadContext.submission(courseID: assignment.courseID, assignmentID: assignment.id)
+        let batchID = "assignment-\(assignment.id)"
+        let attachments = items.map { $0.attachments ?? [] }.reduce([], +)
```

Comment:
I think this could be easier to understand. But I don't know if it's worth being so much longer.
```
var attachments = []
for item in items {
    attachments.append(contentsOf: item.attachments)
}
```